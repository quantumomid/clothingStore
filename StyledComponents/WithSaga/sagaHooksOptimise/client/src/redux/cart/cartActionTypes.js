export const TOGGLE_CART_HIDDEN = "TOGGLE_CART_HIDDEN";
export const ADD_ITEM = "ADD_ITEM";
export const CLEAR_ITEM_FROM_CART = "CLEAR_ITEM_FROM_CART";
export const REMOVE_ITEM = "REMOVE_ITEM";
export const CLEAR_CART = "CLEAR_CART";
export const SET_CART_FROM_FIREBASE = "SET_CART_FROM_FIREBASE";
export const UPDATE_CART_IN_FIREBASE = "UPDATE_CART_IN_FIREBASE";