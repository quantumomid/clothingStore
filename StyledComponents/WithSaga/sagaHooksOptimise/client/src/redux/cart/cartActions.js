import { ADD_ITEM, CLEAR_CART, CLEAR_ITEM_FROM_CART, REMOVE_ITEM, TOGGLE_CART_HIDDEN, UPDATE_CART_IN_FIREBASE, SET_CART_FROM_FIREBASE } from "./cartActionTypes";

export const toggleCartHidden = () => {
    return { type: TOGGLE_CART_HIDDEN }
}

export const addItem = (item) => ({
    type: ADD_ITEM,
    payload: item
})

export const clearItemFromCart = item => ({
    type: CLEAR_ITEM_FROM_CART,
    payload: item
});

export const removeItem = item => ({
    type: REMOVE_ITEM,
    payload: item
});

export const clearCart = () => ({
    type: CLEAR_CART
});

// Needs no argument because updating in FIREBASE not our front end i.e. not redux
export const updateCartItemsInFirebase = () => ({
    type: UPDATE_CART_IN_FIREBASE
});

// Now need to pass cart items as we are setting the front end i.e. redux cart items state!
export const setCartFromFirebase = cartItems => ({
    type: SET_CART_FROM_FIREBASE,
    payload: cartItems
});