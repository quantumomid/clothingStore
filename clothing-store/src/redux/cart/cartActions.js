import { ADD_ITEM, TOGGLE_CART_HIDDEN } from "./cartActionTypes";

export const toggleCartHidden = () => {
    return { type: TOGGLE_CART_HIDDEN }
}

export const addItem = (item) => ({
    type: ADD_ITEM,
    payload: item
})