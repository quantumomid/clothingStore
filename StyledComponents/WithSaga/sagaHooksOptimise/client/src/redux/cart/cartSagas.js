import { takeLatest, put, all, call, select } from "redux-saga/effects";
import { SIGN_OUT_SUCCESS, SIGN_IN_SUCCESS } from "../user/userActionTypes";
import { clearCart, setCartFromFirebase, updateCartItemsInFirebase } from "./cartActions";
import { getUserCartRef } from "../../firebase/firebase.utils";
import { selectCurrentUser } from '../user/userSelectors';
import { selectCartItems } from './cartSelectors';
import { ADD_ITEM, CLEAR_ITEM_FROM_CART, REMOVE_ITEM } from "./cartActionTypes";

export function* clearCartOnSignOut(){
    yield put(clearCart());
}

// This is an example of reusing saga from another slice i.e.
// userSagas here to also perform some logic i.e. clear cart!
export function* onSignOutSuccess(){
    yield takeLatest(SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

// This is run with respect to the SIGN_IN_SUCCESS action type 
// which comes with the user payload (see user actions JS file)
export function* checkCartFromFirebase({ payload: user }) {
    console.log("user in checkCartFromFirebase Saga", user);
    const cartRef = yield getUserCartRef(user.id);
    console.log("user Ref back from getUserCartRef in checkCartFromFirebase Saga", cartRef);
    const cartSnapshot = yield cartRef.get();
    yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
}

// Our whole operation begins open successful sign-in > then only 
// will we check the cart items for the user in firebase
export function* onUserSignIn() {
    yield takeLatest(SIGN_IN_SUCCESS, checkCartFromFirebase);
}

// update the cart items in firebase as the user adds, removes or 
// clears their cart in the front end to keep firebase cart items 
// and redux ones in sync!
export function* updateCartInFirebase() {
    const currentUser = yield select(selectCurrentUser);
    if (currentUser) {
      try {
        const cartRef = yield getUserCartRef(currentUser.id);
        const cartItems = yield select(selectCartItems);
        yield cartRef.update({ cartItems });
        yield put(updateCartItemsInFirebase());
      } catch (error) {
        console.log(error);
      }
    }
}
  
// We also need to track whenever our cart state in redux changes 
// so we can update our carts in firebase ACCORDINGLY!
export function* onCartChange() {
    yield takeLatest(
      [
        ADD_ITEM,
        REMOVE_ITEM,
        CLEAR_ITEM_FROM_CART
      ],
      updateCartInFirebase
    );
}

export function* cartSagas(){
    yield all([call(onSignOutSuccess), call(onUserSignIn), call(onCartChange)]);
}