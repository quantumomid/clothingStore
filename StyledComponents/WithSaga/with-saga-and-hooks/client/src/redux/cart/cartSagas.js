import { takeLatest, put, all, call } from "redux-saga/effects";
import { SIGN_OUT_SUCCESS } from "../user/userActionTypes";
import { clearCart } from "./cartActions";

export function* clearCartOnSignOut(){
    yield put(clearCart());
}

// This is an example of reusing saga from another slice i.e.
// userSagas here to also perform some logic i.e. clear cart!
export function* onSignOutSuccess(){
    yield takeLatest(SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

export function* cartSagas(){
    yield all([call(onSignOutSuccess)]);
}