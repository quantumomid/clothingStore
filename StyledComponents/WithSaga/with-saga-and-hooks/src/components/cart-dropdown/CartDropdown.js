import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom"
import { selectCartItems } from "../../redux/cart/cartSelectors";
import CartItem from "../cart-item/CartItem";
import { toggleCartHidden } from "../../redux/cart/cartActions";
import { CartDropdownButton, CartDropdownContainer, CartItemsContainer, EmptyMessageContainer } from "./cartDropdownStyles";

const CartDropdown = () => {
    const dispatch = useDispatch();
    const cartItems = useSelector(selectCartItems);

    const history = useHistory();
    return (
        <CartDropdownContainer>
            <CartItemsContainer>
            {   cartItems.length ? (
                cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
                : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
            }
            </ CartItemsContainer>
            <CartDropdownButton onClick={() => {
                history.push('/checkout')
                dispatch(toggleCartHidden())   
            }}>GO TO CHECKOUT</CartDropdownButton>
        </ CartDropdownContainer>
    )
};

// Note: the order of wrapping is important i.e. to pass the "connected" component
// into withRouter so that the correct end state of CartDropdown has access to the 
// various router properties we desire
export default CartDropdown;
