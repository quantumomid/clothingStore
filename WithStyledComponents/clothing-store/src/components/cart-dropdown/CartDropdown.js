import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCartItems } from "../../redux/cart/cartSelectors";
import { withRouter } from "react-router";
import CartItem from "../cart-item/CartItem";
import CustomButton from "../custom-button/CustomButton";
// import "./CartDropdown.scss";
import { toggleCartHidden } from "../../redux/cart/cartActions";
import { CartDropdownButton, CartDropdownContainer, CartItemsContainer, EmptyMessageContainer } from "./cartDropdownStyles";

const CartDropdown = ({ cartItems, history, dispatch }) => (
    <CartDropdownContainer>
        <CartItemsContainer>
        {   cartItems.length ? (
            cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
            : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        }
        </ CartItemsContainer>
        <CartDropdownButton onClick={() => {
            history.push('/checkout')
            dispatch(toggleCartHidden())   
        }}>GO TO CHECKOUT</CartDropdownButton>
    </ CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
    cartItems: selectCartItems
});

// Note: the order of wrapping is important i.e. to pass the "connected" component
// into withRouter so that the correct end state of CartDropdown has access to the 
// various router properties we desire
export default withRouter(connect(mapStateToProps)(CartDropdown));

// Also note when the second parameter to connect() is empty then connect automatically
// providers dispatch() in the props for you to use! 