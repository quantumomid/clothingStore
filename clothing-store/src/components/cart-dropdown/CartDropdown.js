import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCartItems } from "../../redux/cart/cartSelectors";
import { withRouter } from "react-router";
import CartItem from "../cart-item/CartItem";
import CustomButton from "../custom-button/CustomButton";
import "./CartDropdown.scss";
import { toggleCartHidden } from "../../redux/cart/cartActions";

const CartDropdown = ({ cartItems, history, dispatch }) => (
    <div className="cart-dropdown">
        <div className="cart-items" >
        {   cartItems.length ? (
            cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
            : <span className="empty-message">Your cart is empty</span>
        }
        </div>
        <CustomButton onClick={() => {
            history.push('/checkout')
            dispatch(toggleCartHidden())   
        }}>GO TO CHECKOUT</CustomButton>
    </div>
);

const mapStateToProps = createStructuredSelector({
    cartItems: selectCartItems
});

// Note: the order of wrapping is important i.e. to pass the "connected" component
// into withRouter so that the correct end state of CartDropdown has access to the 
// various router properties we desire
export default withRouter(connect(mapStateToProps)(CartDropdown));

// Also note when the second parameter to connect() is empty then connect automatically
// providers dispatch() in the props for you to use! 