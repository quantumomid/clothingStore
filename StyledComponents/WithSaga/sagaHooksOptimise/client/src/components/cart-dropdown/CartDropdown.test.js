import { shallow } from "enzyme";
import CartDropdown from "./CartDropdown";
import * as redux from 'react-redux';
import { toggleCartHidden } from '../../redux/cart/cartActions';
import CartItem from "../cart-item/CartItem";

// Mock History parts need to be defined outside describe and 
// test blocks (https://stackoverflow.com/questions/60404319/react-jest-enzyme-testing-usehistory-hook-breaks-test)
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
      push: mockHistoryPush,
    }),
}));

describe("CartDropdown component", () => {
    let wrapper;
    let spyOnUseSelector;
    const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];
    let spyOnUseDispatch;
    let mockDispatch;

    beforeEach(() => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        spyOnUseSelector.mockReturnValue(mockCartItems);
    
        // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);

        wrapper = shallow(<CartDropdown />);
    });

    it("CartDropdown should render", () => {
        expect(wrapper).toMatchSnapshot();
    })

    it('Should call history and dispatch when button is clicked', () => {
        wrapper.find('CartDropdownButton').simulate('click');
        expect(mockHistoryPush).toHaveBeenCalled();
        expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
      });

    it('Should render an equal number of CartItem components as the cartItems redux state', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
    });
})

describe("CartDropdown component when the cart is empty", () => {
    let wrapper;
    let spyOnUseSelector;
    const mockCartItems = [];
    let spyOnUseDispatch;
    let mockDispatch;

    beforeEach(() => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        spyOnUseSelector.mockReturnValue(mockCartItems);
    
        // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);

        wrapper = shallow(<CartDropdown />);
    });

    it("Empty message container to be displayed when cart items are empty", () => {
        expect(wrapper.exists('EmptyMessageContainer')).toBe(true);
    })
})