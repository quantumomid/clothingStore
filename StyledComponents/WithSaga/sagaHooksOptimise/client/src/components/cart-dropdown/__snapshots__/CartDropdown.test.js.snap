// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component CartDropdown should render 1`] = `ShallowWrapper {}`;
